rasa/importers/autoconfig.py

```diff
@@ -0,0 +1,197 @@
+def get_auto_configuration(config_file) -> Dict[Text, Any]:
+    """Determine configuration from a configuration file.
+
+    Keys that are provided in the file are kept. Keys that are not provided are
+    configured automatically.
+    """
+    if config_file and os.path.exists(config_file):
+        config = io_utils.read_config_file(config_file)
+
+        missing_keys = [k for k in CONFIG_AUTOCONFIGURABLE_KEYS if not config.get(k)]
+        config = _create_config_for_keys(config, missing_keys)
+
+        dump_config(config, config_file)
+    else:
+        config = {}
+
+    return config
+
+
+def _create_config_for_keys(
+    config: Dict[Text, Any], keys: List[Text]
+) -> Dict[Text, Any]:
+    """Complete a config by adding automatic configuration for the specified keys.
+
+    Args:
+        config: The provided configuration.
+        keys: Keys to be configured automatically (e.g. `policies`, `pipeline`).
+    """
+    import pkg_resources
+
+    if keys:
+        logger.debug(
+            f"The provided configuration does not contain the key(s) {keys}. "
+            f"Running automatic configuration for them now."
+        )
+
+    default_config_file = pkg_resources.resource_filename(
+        __name__, "default_config.yml"
+    )
+
+    default_config = io_utils.read_config_file(default_config_file)
+
+    config = copy.deepcopy(config)
+    for key in keys:
+        config[key] = default_config[key]
+
+    return config
+
+
+def dump_config(config: Dict[Text, Any], config_file: Text) -> None:
+    """Dump the automatically configured keys into the config file."""
+    import pkg_resources
+
+    try:
+        content = io_utils.read_config_file(config_file)
+    except ValueError:
+        content = ""
+
+    empty = False
+    if not content:
+        empty = True
+        cli_utils.print_warning(
+            f"Configuration file {config_file} does not exist or is empty or invalid. "
+            f"Creating a new one now and filling it with the current configuration."
+        )
+        empty_config_file = pkg_resources.resource_filename(
+            "rasa.cli.initial_project", "config.yml"
+        )
+        content = io_utils.read_config_file(empty_config_file)
+        shutil.copy(empty_config_file, config_file)
+
+    missing_keys = [k for k in CONFIG_KEYS if k not in content.keys()]
+    autoconfigured = [k for k in CONFIG_AUTOCONFIGURABLE_KEYS if not content.get(k)]
+
+    _add_missing_config_keys(missing_keys, config_file)
+
+    lines_to_insert = _get_lines_to_insert(config, autoconfigured)
+
+    try:
+        with open(config_file, "r+", encoding=io_utils.DEFAULT_ENCODING) as f:
+            lines = f.readlines()
+            f.seek(0)
+
+            removing_old_config = False
+            for line in lines:
+                insert_section = None
+
+                if empty and config.get("language") and re.match("language:", line):
+                    line = f"language: {config['language']}\n"
```

Comment:
It's safer to replace it.

`empty` is only `True` if something somehow went wrong and the config file disappeared or got replaced while the auto configuration was running: In the beginning of `get_configuration`, we check `if config_file and os.path.exists(config_file)`, and only then do we try to dump later on. If this is during training, we've also [validated already](https://github.com/RasaHQ/rasa/blob/2c97e1c5d8b0ca02b2c9438cebf74bdabdf6f3b7/rasa/cli/train.py#L61) that the file has some content (it needs to contain at least the key `language`).

So if we still end up in this if-branch, something has gone wrong, and we would like to just dump the current configuration. (**Or maybe we shouldn't do that and just throw an error instead?**) To do so, we load an empty one and fill `policies` and `pipeline` with the current configuration. Now we also need to replace `language` with the current language configuration, otherwise it's always going to be the default `en`.

Reply:
> (Or maybe we shouldn't do that and just throw an error instead?

I'd actually prefer this one as this case (file moved, somehow got unaccessible while auto configuring) should be super rare and I'd assume that we might then also run in problems with dumping

Reply:
I changed the structure a little bit:

In the beginning of `dump_config`, we now check if the config file looks as it should, i.e.
- exists
- not empty
- missing the same keys it was missing when we first checked in `get_configuration`
- no configuration specified for the same keys as when we first checked in `get_configuration`

If any of these is not the case, that means something went wrong - the config file shouldn't be modified or removed while we run auto configuration - and we throw an error and don't dump anything.